### PR TITLE
Add Model Scanning

### PR DESCRIPTION
**Pull request by blessedcoolant**
Uses the [picklescan](https://github.com/mmaitre314/picklescan) library to scan the models that are being loaded to InvokeAI.

If the model is infected in anyway, the app shuts down immediately and the user is warned. If the model is unable to be scanned, then the user is presented with options to either continue or exit.

### CODE REVIEW

File: installer/requirements.in [hunk 1]
```diff
@@ -19,6 +19,7 @@ torch-fidelity
 torchvision==0.13.1 ; platform_system == 'Darwin'
 torchvision==0.13.1+cu116 ; platform_system == 'Linux' or platform_system == 'Windows'
 transformers
+picklescan==0.0.5
```
Comment (tildebyte) [inline installer/requirements.in:22]: We shouldn't pin this in here (let pip-compile figure it out), UNLESS there's some hard requirement to do so

File: ldm/invoke/model_cache.py [hunk 2]
```diff
@@ -275,6 +278,32 @@ def offload_model(self, model_name:str):
         gc.collect()
         if self._has_cuda():
             torch.cuda.empty_cache()
+    
+    def scan_model(self, model_name, checkpoint):
+        # scan model
+        from picklescan.scanner import scan_file_path
```
Comment (tildebyte) [inline ldm/invoke/model_cache.py:284]: Can these imports move to the top of this file?

File: ldm/invoke/model_cache.py [hunk 3]
```diff
@@ -275,6 +278,32 @@ def offload_model(self, model_name:str):
...
+        import sys
```
Comment (tildebyte) [inline ldm/invoke/model_cache.py:285]: Can these imports move to the top of this file?

**Comment by tildebyte** [comment]
LGTM, but I'll wait for@lstein to test this out and ACK or not

**Review by lstein — CHANGES_REQUESTED** [review]
I think we need changes to the environments files as well.

**Comment by blessedcoolant** [comment]
> I think we need changes to the environments files as well.

Done.

**Comment by lstein** [comment]
Tested this out on Linux. It didn't crash, the scan was fast, and reported OK. Have you tested a corrupted or malicious .ckpt file and know that the scanner will catch problems?

Please update requirements in the environments files, and this is good to go, I think.

**Review by lstein — APPROVED** [review]
Looking good.

**Comment by blessedcoolant** [comment]
> Tested this out on Linux. It didn't crash, the scan was fast, and reported OK. Have you tested a corrupted or malicious .ckpt file and know that the scanner will catch problems?

I haven't tested it with a corrupt file coz I didn't know where to find one. However the library seems to be popular and well made. @tildebyte might have tested / heard of it working correctly.

I had a look at the library itself. It basically looks for any eval calls -- which is how most malicious code would be executed. As long as it does that effectively, I think this'll do the job.

From our end, this is not meant to be a hardcore safety check but another line of defense that will help users. Which I feel it does.

**Comment by McPatate** [comment]
Hey, thanks for using picklescan !

Imho it should cover your needs for the moment. We use it on the hugging face hub to get the imports list you see in our 'Files and versions' tab.

You can see an example of malicious pickle [here](https://huggingface.co/ykilcher/totally-harmless-model/tree/main).

**Comment by blessedcoolant** [comment]
> Hey, thanks for using picklescan !
> 
> Imho it should cover your needs for the moment. We use it on the hugging face hub to get the imports list you see in our 'Files and versions' tab.
> 
> You can see an example of malicious pickle [here](https://huggingface.co/ykilcher/totally-harmless-model/tree/main).

Hey, thanks for reaching out. Yeah, it seems like it'll get the job done. And it is inactive development so I think we should be set for a while into the future too.

**Pull request merged**